chapters/01.Brackets/brackets.jl

```diff
@@ -0,0 +1,1156 @@
+### A Pluto.jl notebook ###
+# v0.16.0
+
+using Markdown
+using InteractiveUtils
+
+# ╔═╡ 9c340f8b-877d-41b9-baca-2698cc2b5988
+using Plots
+
+# ╔═╡ 99adb0e2-1b94-11ec-1351-79707f6016af
+md"""
+# Bracket search
+
+*STMO*
+
+**Michiel Stock**
+"""
+
+# ╔═╡ 6b30fe20-f321-4cb8-a2fc-0a2820d71bc0
+md"As a first introduction to mathematical optimization, we will study two *bracketing methods*. Bracketing methods can be used to minimize scalar function with a single input variable. These algorithms identify an interval $[a, b]$ containing the desired minimum. We will assume that the functions are *unimodal* they contain a single minimum."
+
+# ╔═╡ bdf6c059-db72-4d90-b98f-2e77b7d89f37
+md"""
+## Finding an initial bracket
+
+The following function will generate from an initial $x$ an interval containing a local minimum. It works by moving one of the limits of the interval until the function value at that limit starts to decrease.
+"""
+
+# ╔═╡ 662b4d13-505c-4202-b104-0db80d8b1822
+function bracket_minimum(f, x=0.0; s=1e-2, k=2.0)
+  a, ya = x, f(x)  # lower limit of the interval
+  b, yb = x, f(x + s)  #  upper limit of the interval
+  if yb > ya  # default assumes f is decreasing in a, flip if not the case
+    a, b = b, a
+    ya, yb = yb, ya
+    s = -s  # go to the left instead of right
+  end
+  while true
+    c, yc = b + s, f(b + s)
+    if yc > yb
+      return a < c ? (a, c) : (c, a)
+    end
+    a, ya = b, yb
+    b, yb = c, yc
+    s *= k
+  end
+end
+
+# ╔═╡ 42131c7f-2980-4bc9-be00-64e25ecd238e
+
+
+# ╔═╡ 143801cb-25ee-4a49-882a-a8ab377a25f8
+md"""
+## Bisection method
+
+The bisection method is technically not a minimization algorithm, but a *root finding method*, i.e., it can be used to find an $x$ for a given function $g: \mathbb{R}\rightarrow \mathbb{R}$ such that
+
+$$g(x) = 0\,.$$
+
+We can use the bisection method to find a minimum of a function $f(x)$ by finding points where the first derivative $f'(x)$ is equal to 0.
+
+> **Question**: in addition to the first derivative being equal to zero, which other criteria should hold for $x^\star$ to be a minimizer?
+
+The bisection method departs from an initial bracket $[a, b]$, where $g(a)$ and $g(b)$ have opposing signs. If $g(x)$ is a continious function, the *intermediate value theorem* states that there is at least one $x\in[a,b]$ such that $g(x)$.
```

Comment:
✏️ If $g(x)$ is a **continuous** function, the intermediate value theorem states that there is at least one $x\in[a,b]$ such that $g(x)$ **= 0**.